scio-core/src/main/scala/com/spotify/scio/values/ApproxFilter.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.values
+import java.io._
+
+import com.google.common.io.{ByteStreams, CountingOutputStream}
+import com.spotify.scio.annotations.experimental
+import com.spotify.scio.coders.Coder
+import org.apache.beam.sdk.coders.AtomicCoder
+
+/**
+ * An [[ApproxFilter]] is an abstraction over various Approximate / Probabilistic
+ * data structures used for checking membership of elements.
+ *
+ * This trait defines read-only immutable filters. The filters are primarily aimed
+ * to be used as singleton [[SideInput]]s in Scio pipelines.
+ *
+ * For example usage see [[BloomFilter]]
+ */
+@experimental
+trait ApproxFilter[-T] extends Serializable {
```

Comment:
Is variance really necessary? If it's gonna be used in an `SCollection[T]`, `T` is always invariant.

Reply:
I am not sure if this needs to be related with invariance of an `SCollection[T]`. This is because the types `ApproxFilter` and `SCollection` are never related.

Since `ApproxFilter` is read only and meant to be used as a predicate, the type is contravariant.
When casted to an `ApproxFilter` the contravariant type restricts users from building monoids.
Eg: you cannot have a `putAll( other: ApproxFilterImpl[T]): ApproxFilterImpl[T]` if it is contravariant.

My original intention was to easily support reusing `BloomFilterSideInputs` to do `sparseIntersectByKey` or filtering multiple collections without having to build the bloom filters multiple times, and also keeping the impl as generic as we can. But now that I tried this out, it doesn't need to be contravariant, however the type signatures are a bit complicated.
```
  sealed trait Animal
  case class Cat() extends Animal
  case class Dog() extends Animal
  implicit def f: Funnel[Animal] = ???

  val animals: SCollection[Animal] = ???
  val cats: SCollection[Cat] = ???
  
  val animalsBf = animals.toBloomFilter(0.01)
  cats.appoxFilter(animalsBf)
```

The types end up like this:
```
def appoxFilter[S >: T: Coder, AF[_] <: ApproxFilter[_]](filter: SCollection[AF[S]])(implicit afcoder: Coder[AF[S]]): SCollection[T] = {
```

I'll park this this to implement later and see where we end up after a few iterations.

Reply:
Yeah but how common are these cases, when key is not a simple string or case class?